pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/NumericalFilterOptimizer.java

```diff
@@ -271,7 +242,211 @@ private Expression rewrite(Expression equals, Expression lhs, Expression rhs, Sc
             break;
           }
         }
+      }
     }
     return equals;
   }
+
+  /**
+   * Rewrite expressions of form "column > literal", "column >= literal", "column < literal", and "column <= literal"
+   * to ensure that RHS literal is the same datatype as LHS column.
+   */
+  private static Expression rewriteRangeExpression(Expression range, FilterKind kind, Expression lhs, Expression rhs, Schema schema) {
+    // Get column data type.
+    FieldSpec.DataType dataType = schema.getFieldSpecFor(lhs.getIdentifier().getName()).getDataType();
+
+    switch (rhs.getLiteral().getSetField()) {
+      case SHORT_VALUE:
+      case INT_VALUE:
+        // No rewrites needed since SHORT and INT conversion to numeric column types (INT, LONG, FLOAT, and DOUBLE) is
+        // lossless and will be implicitly handled on the server side.
+        break;
+      case LONG_VALUE: {
+        long actual = rhs.getLiteral().getLongValue();
+        switch (dataType) {
+          case INT: {
+            int converted = (int) actual;
+            int comparison = Long.compare(actual, converted);
+            if (comparison > 0) {
+              // Literal value is greater than the bounds of INT. > and >= expressions will always be false because an
+              // INT column can never have a value greater than Integer.MAX_VALUE. < and <= expressions will always be
+              // true, because an INT column will always have values less than Integer.MIN_VALUE.
+              setExpressionToBoolean(range, kind == FilterKind.LESS_THAN || kind == FilterKind.LESS_THAN_OR_EQUAL);
+            } else if (comparison < 0) {
+              // Literal value is less than the bounds of INT. > and >= expressions will always be true because an
+              // INT column will always have a value greater than or equal to Integer.MIN_VALUE. < and <= expressions
+              // will always be false, because an INT column will never have values less than Integer.MIN_VALUE.
+              setExpressionToBoolean(range, kind == FilterKind.GREATER_THAN || kind == FilterKind.GREATER_THAN_OR_EQUAL);
+            } else {
+              // Long literal value falls within the bounds of INT column, server will successfully convert the literal
+              // value when needed.
+            }
+            break;
+          }
+          case FLOAT: {
+            // Since we are converting a long value to float, float value will never be out of bounds (i.e -Infinity
+            // or +Infinity).
+            float converted = (float) actual;
```

Comment:
How is this any different from the case for INT above where the LONG is being converted to INT

Reply:
When an out of bounds value is converted to int, the int value becomes either Integer.MIN_VALUE or Integer.MAX_VALUE. However, when an out of bounds value is converted to float, the the float value becomes -Infinity or +Infinity. The code difference is mainly because of this.